packages/client/wallets/smart-wallet/src/blockchain/wallets/SendTransactionService.ts

```diff
@@ -0,0 +1,195 @@
+    constructor(
+        message: string,
+        public readonly viemError: BaseError,
+        code = WalletErrorCode.SEND_TRANSACTION_CONFIRMATION_FAILED
+    ) {
+        super(message, viemError, code);
+    }
+}
+
+/**
+ * Error thrown when a transaction simulation fails.
+ * @param viemError The error thrown by the viem client. See https://viem.sh/docs/glossary/errors.html
+ */
+export class EVMSendTransactionSimulationError extends EVMSendTransactionError {
+    constructor(
+        message: string,
+        public readonly viemError: BaseError,
+        code = WalletErrorCode.SEND_TRANSACTION_SIMULATION_FAILED
+    ) {
+        super(message, viemError, code);
+    }
+}
+
+/**
+ * Error thrown when a transaction fails due to a contract execution error.
+ * @param viemError The error thrown by the viem client. See https://viem.sh/docs/glossary/errors.html
+ * @param revertError The revert error from viem containing the reason for the revert.
+ * @param revertError.reason The reason for the revert.
+ * @param revertError.data The decoded revert error data.
+ * @example
+ * try {
+ *   await wallet.sendTransaction({
+ *     address: contractAddress,
+ *     abi,
+ *     functionName: "mintNFT",
+ *     args: [recipientAddress],
+ *   });
+ * } catch (e) {
+ *   if (e instanceof SendTransactionExecutionRevertedError) {
+ *     alert(`Transaction reverted: ${e.revertError.reason}`);
+ *   }
+ *   throw e;
+ * }
+ */
+export class EVMSendTransactionExecutionRevertedError extends EVMSendTransactionSimulationError {
+    constructor(
+        message: string,
+        public readonly viemError: BaseError,
+        public readonly revertError: ContractFunctionRevertedError,
+        public txId: string | undefined
+    ) {
+        super(message, viemError, WalletErrorCode.SEND_TRANSACTION_EXECUTION_REVERTED);
+    }
+}
+
+export interface SendTransactionOptions {
+    confirmations: number;
+    transactionConfirmationTimeout: number;
+    awaitConfirmation?: boolean;
+}
+
+export class SendTransactionService {
+    constructor(
+        private publicClient: PublicClient,
+        private defaultSendTransactionOptions: SendTransactionOptions = {
+            confirmations: 2,
+            transactionConfirmationTimeout: 30_000,
+            awaitConfirmation: true,
+        }
+    ) {}
+
+    async sendTransaction(
+        request: TransactionServiceTransactionRequest,
+        client: SmartAccountClient<EntryPoint, Transport, Chain, SmartAccount<EntryPoint>>,
+        config: Partial<SendTransactionOptions> = {}
+    ): Promise<Hex> {
+        const { confirmations, transactionConfirmationTimeout, awaitConfirmation } = this.getConfig(config);
+        try {
+            await this.simulateCall(request, undefined);
+            const hash = await client.writeContract({
+                ...request,
+                account: client.account,
+                chain: client.chain,
+            });
+            if (awaitConfirmation) {
```

Comment:
If there's any other config we never actually use I'd also consider dropping!

Reply:
Alfonso asked for it to be included

Reply:
We're not exposing it through the API though

Reply:
So seems like we should either delete the functionality, or expose the option through `callContract`

@alfonso-paella I think we should skip the confirmations for now, and wait until someone asks for it. I suspect anyone who's sophisticated enough to care will use the viem clients.

Also, it's not a breaking change to add this later. The reverse is not true.

Reply:
Whoops you're right. @alfonso-paella I'd tend to agree that we shouldn't expose the option to not wait for confirmations, anyone who is just interested in landing a transaction from a smart wallet would want it enabled and anyone doing something that would require disabling it (e.g. their own transaction management logic) would almost certainly be using the viem client

Reply:
I forgot but I think what I'd suggested in the end is we allow a customizable number of blocks, not that we allow turning awaiting confirmation off

Can we do that?

Because some use cases may prefer low latency, others may prefer low chance of reorg

Is that hard to do?

Reply:
I added two options to the SDK call:
- Transaction confirmations
- Confirmation timeout (how long to wait before throwing an error)